### DISCUSSION TITLE
What is best practice to dynamically CRUD columns in igx-grid

### DISCUSSION THREAD

**Discussion by KaushikEDS** [IgniteUI/igniteui-angular#13960] [Q&A] [answered]
I have a use case where user can manually add which columns he wants in the grid.
I want the best possible solution with state management (using Ngrx) where all interactions with grid are as smooth as possible.
The dynamic columns can have filters, sorting, grouping, summaries, etc.

Also the type of columns will be different, so want the best way to apply formatting. For example if column shows price of a Stock it should have $ sign with it.

User can save the grid and on refresh the same grid should be visible.
https://www.infragistics.com/products/ignite-ui-angular/angular/components/grid/state-persistence
I'm thinking that will be able to help.

I would like an example or links to the code snippets/documentation if possible.

**Comment by RivaIvanova** [comment, marked as answer]
Hello,

I have been looking into your questions and here are some directions regarding the different points:

> I have a use case where user can manually add which columns he wants in the grid.

I am not quite sure what you mean by _"manually add which columns he wants in the grid"_, however, if you require showing/hiding some columns, I believe that you will find our **Column Hiding** feature topic [here](https://www.infragistics.com/products/ignite-ui-angular/angular/components/grid/column-hiding) quite helpful on the matter. Additionally, you could also set the `IgxColumn`’s [hidden](https://www.infragistics.com/products/ignite-ui-angular/docs/typescript/latest/classes/IgxColumnComponent.html#hidden) property in order to hide/show it.

> I want the best possible solution with state management (using Ngrx) where all interactions with grid are as smooth as possible.


The NgRx framework is not directly related to the Ignite UI for Angular components and as such it could be considered beyond our scope. I will suggest referring to the official NgRx documentation [here](https://ngrx.io/guide/store) where you could further investigate the requirements and setup.


> The dynamic columns can have filters, sorting, grouping, summaries, etc.



In order to define the `IgxColumn`, you could achieve this by defining the columns collection in the markup as mentioned in our **Defining columns** section [here](https://www.infragistics.com/products/ignite-ui-angular/angular/components/grid/grid#defining-columns). Additionally, if these columns are dynamic, you could define them by adding the `<igx-column>` tag in the template and using `*ngFor="let col of columns` in order to add as many columns as you want in the IgxGrid.

For example:

**component.html**
```html
<igx-grid #grid [data]="data" [height]="'600px'" [allowFiltering]="true" [filterMode]="'quickFilter'">
  <igx-column
    *ngFor="let c of columns"
    [sortable]="true"
    [filterable]="true"
    [field]="c.field"
    [header]="c.header"
    [dataType]="c.dataType"
    [pinned]="c.pinned"
    [filterable]="c.filterable"
    [sortable]="c.sortable"
    [groupable]="c.groupable"
    [hasSummary]="c.hasSummary"
    [cellTemplate]="c.cellTemplate"
  >
  </igx-column>
</igx-grid>
```
**component.ts**
```ts
this.columns = [
  {
    field: 'ID',
    header: 'ID',
    width: '100px',
    dataType: 'number',
    pinned: true,
    filterable: true,
    sortable: true,
    groupable: false,
    hasSummary: true,
  },
  ...
];
```
[Here](https://stackblitz.com/edit/q299rp?file=src%2Fapp%2Fgrid%2Fgrid-boston-marathon%2Fgrid.component.html,src%2Fapp%2Fgrid%2Fgrid-boston-marathon%2Fgrid.component.ts) could be found a small sample demonstrating my suggestion.


> Also the type of columns will be different, so want the best way to apply formatting. For example if column shows price of a Stock it should have $ sign with it.


I would suggest checking out our **Grid Column Types** topic [here](https://www.infragistics.com/products/ignite-ui-angular/angular/components/grid/column-types) as it provides detailed information along with code snippets and a sample demonstrating how to enable a data type-specific template. 


> User can save the grid and on refresh the same grid should be visible.
> https://www.infragistics.com/products/ignite-ui-angular/angular/components/grid/state-persistence
> I'm thinking that will be able to help.


Yes, the `IgxGridState` directive is the correct feature to use in order to save and restore the `IgxGrid` state with many supported features such as Sorting, Filtering, Advanced Filtering, Paging, Cell Selection, Row Selection, Column Selection, Row Pinning, Expansion, GroupBy, Columns.

[Here](https://www.infragistics.com/products/ignite-ui-angular/angular/components/grid/state-persistence#demo) could be found a sample demonstrating how the state persistence could be configured.

I hope the provided resources help you achieve your requirements.
